SPIKE.md

```diff
@@ -0,0 +1,92 @@
+# Spike: Robust handling of ObjectID type for MongoDB
+
+Ref: https://github.com/strongloop/loopback-next/issues/3456
+
+## Table of contents
+
+- [The problem](#the-problem)
+- [Proposed solution](#proposed-solution)
+- [Approaches](#approaches)
+- [Developer Notes](#developer-notes)
+- [Implementation Proposal](#implementation)
+- [Follow-up task](#follow-up-task)
+
+## The Problem
+
+`ObjectID` is a native MongoDB datatype, which is used as the datatype of its primary key (_id_) field; other fields
+can also be of `ObjectID` type.
+
+In the current version of Juggler, the interpretation of `ObjectID` field is a sub-optimal experience supported by a complex
+determination process in an already complex codebase.
+
+Examples of bad experiences: 
+
+1. Fields need not be set to `mongodb: {dataType: 'ObjectID'}` to be interpreted as `ObjectID`.
+2. Unless `strictObjectIDCoercion` is set to `true`, any string
+that looks like an `ObjectID` will automatically get converted to `ObjectID` and you won't be able to find the object if you
+tried searching the database by that string, because it is not a string any more.
+3. To add to the confusion, this behavior can be configured at model and property levels.
+
+Complexity is sometimes unavoidable, but a less complex state is always the desired state. It creates better experience for
+the consumers, and makes development and maintenance a more pleasant experience for the creators.
+
+## Proposed Solution
+
+"Do not interpret any property not set with `mongodb: {dataType: 'ObjectID'}` as an `ObjectID`. Use string values for `ObjectID`
+in public APIs and manage the coercion processes for the user."
```

Comment:
> Do not interpret any property not set

In my experience, double negatives are difficult to understand, especially for native English speaker. Let's rephrase this in a positive way (what we want, instead what we do not want).


```suggestion
Database-specific types like `ObjectID` and `Decimal128` should be considered as an implementation detail of a database connector. Values of these types must not leak from the connector to juggler and user code, code outside the connector should see a database-agnostic type like a `string`. The connector must always convert incoming values (property values, filtering queries) from database-agnostic type to database-specific type on input, and convert database-specific types back to database-agnostic type on output.

Specifically for `ObjectID`:

- Properties that should be stored as `ObjectID` in MongoDB must be defined as `{type: 'string', mongodb: {dataType: 'ObjectID'}}`.
- There will be no auto-magic coercion from `string` to `Object` for other properties (those defined without `dataType: 'ObjectID'`).
```